src/Engine/GameAction.java

```diff
@@ -828,6 +831,100 @@ public UnitActionType getUnitActionType()
     }
   } // ~ResupplyAction
 
+  // ===========  RepairUnitAction  ===============================
+  public static class RepairUnitAction implements GameAction
+  {
+    private Path movePath;
+    private XYCoord startCoord;
+    private XYCoord moveCoord;
+    private XYCoord repairCoord;
+
+    public RepairUnitAction(XYCoord startLoc, Path path, int targetX, int targetY)
+    {
+      this(startLoc, path, new XYCoord(targetX, targetY));
+    }
+
+    public RepairUnitAction(XYCoord startLoc, Path path, XYCoord repairLoc)
+    {
+      movePath = path;
+      startCoord = startLoc;
+      repairCoord = repairLoc;
+      if( null != path && (path.getEnd() != null) )
+      {
+        moveCoord = new XYCoord(movePath.getEnd().x, movePath.getEnd().y);
+      }
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster gameMap)
+    {
+      // Repair actions consist of
+      //   MOVE
+      //   HEAL
+      //   REPAIR
+      GameEventQueue repairEvents = new GameEventQueue();
+
+      boolean isValid = true;
+      Unit benefactor = null;
+      Unit beneficiary = null;
+
+      if( (null != gameMap) && (null != startCoord) && (null != repairCoord) &&
+          gameMap.isLocationValid(startCoord) && gameMap.isLocationValid(repairCoord) )
+      {
+        benefactor = gameMap.getLocation(startCoord).getResident();
+        beneficiary = gameMap.getLocation(repairCoord).getResident();
+        isValid &= benefactor != null && !benefactor.isTurnOver;
+        isValid &= (null != beneficiary) && !benefactor.CO.isEnemy(beneficiary.CO);
+        isValid &= (null != gameMap) && (gameMap.isLocationValid(repairCoord));
+        isValid &= (movePath != null) && (movePath.getPathLength() > 0);
+      }
+      else
+        isValid = false;
+
+      if( isValid )
+      {
+        Location moveLocation = gameMap.getLocation(moveCoord);
+        isValid &= (null == moveLocation.getResident()) || (benefactor == moveLocation.getResident());
+      }
+
+      if( isValid )
+      {
+        if( Utils.enqueueMoveEvent(gameMap, benefactor, movePath, repairEvents) )
+        {
+          // No surprises in the fog.
+          repairEvents.add(new HealUnitEvent(beneficiary, true));
+          repairEvents.add(new ResupplyEvent(beneficiary));
+        }
+      }
+      return repairEvents;
+    }
+
+    @Override
+    public XYCoord getMoveLocation()
+    {
+      return moveCoord;
+    }
+
+    @Override
+    public XYCoord getTargetLocation()
+    {
+      return repairCoord;
+    }
+
+    @Override
+    public String toString()
+    {
+      return String.format("[Heal at %s after moving to %s]",
```

Comment:
Howsabout "Heal %s at %s after moving to %s"?

Reply:
"Heal unit at %s from %s after moving to %s"?

Reply:
The interacting units aren't available for... *reasons*.

Reply:
"That thing that will never happen"?

Currently, Attack, Capture, Wait, Load, Unload, Join, Resupply, Transform, Explode, and Delete track the interacting unit(s). So, every action type except this one and Production.

"Heal Tank at (3, 4) after moving to (3, 3)" seems better than
"Heal at (3, 4) after moving to (3, 3)" or
"Heal unit at (3, 4) from (3, 3) after moving to (3, 3)"

Reply:
Netplay.
It's occurred to me since that it's desirable even if we never get netplay, for replays.
We could just store the starting gamestate and play back the GameActions one by one, with an "AI" that just polls the queue of available actions.

I'm thinking we don't want to store the units, because the units will be instantiated on the fly and will have different IDs than the serialized events... Unless maybe the build action kept a reference to the unit it will produce? (Seems hokey to me)

That logging is nicer, yes. We could probably compromise on it and make the unit references transient... but then the thing you're logging might end up uninitialized after deserialization.


For reference, I think replays are the next feature on my general radar, since that'll be really nice for both "spectators" and the actual players to know what's going on.

I've been thinking the "AI that just pipes out GameActions in the right order" is the way to go here, but I've been looking at some [interesting](https://randomascii.wordpress.com/2013/07/16/floating-point-determinism/) [articles](http://yosefk.com/blog/consistency-how-to-defeat-the-purpose-of-ieee-floating-point.html) on this sort of "lockstep" behavior, and it appears floating-point math is liable to mess everything up without a lot of work. Perhaps just move to using integer percents everywhere?

Another approach would be to store the events themselves. I didn't like that idea because the events just can't get away from storing the interacting units, but it probably wouldn't be a problem if, say, the Create event actually kept a reference to the unit it already built and only ever instantiated it once.
This bears a certain amount of cost in that we'll have to make sure all of the events daisy-chain properly after serializing/deserializing.
This *also* has two advantages:
There's no new "AI" to build, you just load the events in the queue and off you go (assuming we can trigger "end turn" from there)
The lack of floating point determinism shouldn't be an issue, if we avoid events-spawning-events, since unit death will be 100% determined by "is there a death event in the queue?". This probably works if we just always require insertion into the event queue before execution and save events off as they're executed.

Reply:
All right. Well. I haven't properly ruminated on all this yet, but my gut reflex is that we do it one way for 90% of cases, so we should probably follow suit in this one. Any change of this sort is gonna be large anyway, and it will be simple to change this one with all the others when we figure out how to do it.
I'd rather have the more informative message for the time in between now and whenever we decide and figure out how to change things in 2025.

Reply:
Also at one point I wanted to figure out how to make our AI deterministic, but I didn't go to far because then all the AI matches would be fixed. Still not sure why it isn't deterministic - maybe the natural ordering of units in the queue varies from run to run or something. Have to figure that out if we want replays to be a thing.

Reply:
> All right. Well. I haven't properly ruminated on all this yet, but my gut reflex is that we do it one way for 90% of cases, so we should probably follow suit in this one. Any change of this sort is gonna be large anyway, and it will be simple to change this one with all the others when we figure out how to do it.

That's probably fair.

> I'd rather have the more informative message for the time in between now and whenever we decide and figure out how to change things in 2025.

I did say replays was my next big feature goal (though destructible terrain may take precedence, we'll see).
I suppose it makes sense to use the current scheme, since it's nicer right now and we're not sure which way the replays should be set up yet anyhow.

> Also at one point I wanted to figure out how to make our AI deterministic, but I didn't go to far because then all the AI matches would be fixed.

IDK, it'd make debugging issues with them a lot easier.

> Still not sure why it isn't deterministic - maybe the natural ordering of units in the queue varies from run to run or something.

Pretty sure that's it exactly. The coordinate and/or location and/or unit IDs won't all be the same from run to run, so builds and units aren't considered in exactly the same order each time. To fix it, we'd need to make sure to use deterministic-order collections.
...which may or may not be annoying to do, depending on how ties are broken. Could maybe just have an "Age" var on everything to do that, with instantiation order determining it and being deterministic?

> Have to figure that out if we want replays to be a thing (unless we just store all actions directly instead of re-running the AI, of course).

Just storing the actions is exactly what I had in mind. IMO it's totally valid to use RNG in the AI, since the AI's calculations should only ever be done once for any given game.
(In my vague ideal end-state, there'd be some way to add new AI at run-time to allow stuff like the Starcraft AI tournaments, so I really don't want to constrain AI behavior.)

Reply:
Yeah, I'm not gonna try to bite off plugin-based AI modules just yet.

Still want to change this action message though, even if it's "temporary".